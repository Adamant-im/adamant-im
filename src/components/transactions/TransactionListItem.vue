<template>
  <v-list-item>
    <template #prepend>
      <v-list-item-title :class="classes.title">
        <slot name="prepend" />
        {{ title }}
        <slot name="append" />
      </v-list-item-title>
    </template>

    <v-list-item-title v-if="!hidden" :class="classes.value">
      <slot />
    </v-list-item-title>
  </v-list-item>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

const className = 'transaction-list-item'
const classes = {
  root: className,
  title: `${className}__title`,
  value: `${className}__value`
}

export default defineComponent({
  props: {
    title: {
      type: String,
      required: true
    },
    hidden: {
      type: Boolean
    }
  },
  setup() {
    return {
      classes
    }
  }
})
</script>

<style lang="scss" scoped>
@use 'sass:map';
@use '@/assets/styles/settings/_colors.scss';

.transaction-list-item {
  &__title {
    font-weight: 300;
  }
  &__value {
    font-weight: 300;
    font-size: 14px;
    text-align: right;
    text-overflow: ellipsis;
    overflow: hidden;
    max-width: 100%;
    width: 100%;
  }
}

.v-theme--light {
  .transaction-list-item {
    &__title {
      color: map.get(colors.$adm-colors, 'regular');
    }
    &__value {
      color: map.get(colors.$adm-colors, 'muted') !important;
    }
  }
}

.v-theme--dark {
  .transaction-list-item {
  }
}
</style>
