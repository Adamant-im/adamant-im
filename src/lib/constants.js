export const EPOCH = Date.UTC(2017, 8, 2, 17, 0, 0, 0)

export const Transactions = {
  SEND: 0,
  SIGNATURE: 1,
  DELEGATE: 2,
  VOTE: 3,
  MULTI: 4,
  DAPP: 5,
  IN_TRANSFER: 6,
  OUT_TRANSFER: 7,
  CHAT_MESSAGE: 8,
  STATE: 9
}

export const Cryptos = {
  ADM: 'ADM',
  ETH: 'ETH',
  BNB: 'BNB',
  DOGE: 'DOGE',
  BZ: 'BZ'
}

export const ERC20 = Object.freeze([
  Cryptos.BNB,
  Cryptos.BZ
])

export const isErc20 = crypto => ERC20.includes(crypto)

/** Number of decimal places for the different crypto amounts */
export const CryptoAmountPrecision = {
  ADM: 2,
  ETH: 6,
  BNB: 6,
  DOGE: 8,
  BZ: 6
}

/** Fees for the misc ADM operations */
export const Fees = {
  /** Storing a value into the KVS */
  KVS: 0.001,
  /** Transfering tokens */
  ADM_TRANSFER: 0.5,
  NOT_ADM_TRANSFER: 0.001
}

/** Regex for detecting of base64 encoded string */
export const base64regex = /^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)$/

export const Symbols = {
  CLOCK: String.fromCharCode(0x23f0), // ⏰
  HOURGLASS: String.fromCharCode(0x23f3), // ⏳
  CROSS: String.fromCharCode(0x274c) // ❌
}

export const Delegates = {
  ACTIVE_DELEGATES: 101
}

/** Gas value for the ETH transfers */
export const ETH_TRANSFER_GAS = 21000

/** Gas value for the ERC-20 transfers */
export const ERC20_TRANSFER_GAS = ETH_TRANSFER_GAS * 10

export default {
  EPOCH,
  Transactions
}

export const UserPasswordArticleLink = 'https://medium.com/adamant-im/more-convenience-login-to-the-web-messenger-with-user-password-9d48a736dfd8'

export const UserPasswordHashSettings = {
  SALT: 'salt',
  ITERATIONS: 100000,
  KEYLEN: 64,
  DIGEST: 'sha512'
}
