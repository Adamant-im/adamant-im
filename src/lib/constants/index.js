import { Cryptos, CryptosInfo, CryptosOrder } from './cryptos'

export const EPOCH = Date.UTC(2017, 8, 2, 17, 0, 0, 0)

export const Transactions = {
  SEND: 0,
  SIGNATURE: 1,
  DELEGATE: 2,
  VOTE: 3,
  MULTI: 4,
  DAPP: 5,
  IN_TRANSFER: 6,
  OUT_TRANSFER: 7,
  CHAT_MESSAGE: 8,
  STATE: 9
}

/**
 * @see https://github.com/Adamant-im/adamant/wiki/Message-Types
 */
export const MessageType = {
  BASIC_ENCRYPTED_MESSAGE: 1,
  RICH_CONTENT_MESSAGE: 2,
  cryptoTransferMessage: (cryptoSymbol) => `${cryptoSymbol.toLowerCase()}_transaction`
}

export const Rates = {
  USD: 'USD',
  EUR: 'EUR',
  RUB: 'RUB',
  CNY: 'CNY',
  JPY: 'JPY'
}

export const RatesNames = {
  [Rates.USD]: 'USD ($)', // United States Dollar
  [Rates.EUR]: 'EUR (‚Ç¨)', // Euro
  [Rates.RUB]: 'RUB (‚ÇΩ)', // Russian Ruble
  [Rates.CNY]: 'CNY (¬•)', // Chinese Yuan
  [Rates.JPY]: 'JPY (¬•)' // Japanese Yen
}

/** Fees for the misc ADM operations */
export const Fees = {
  /** Storing a value into the KVS */
  KVS: 0.001,
  /** Transferring tokens */
  ADM_TRANSFER: 0.5,
  NOT_ADM_TRANSFER: 0.001
}

/** Regex for detecting of base64 encoded string */
export const base64regex =
  /^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)$/

export const Symbols = {
  CLOCK: String.fromCharCode(0x23f0), // ‚è∞
  HOURGLASS: String.fromCharCode(0x23f3), // ‚è≥
  CROSS: String.fromCharCode(0x274c) // ‚ùå
}

export const Delegates = {
  ACTIVE_DELEGATES: 101
}

export const WelcomeMessage = {
  ADAMANT_BOUNTY: 'ADAMANT Bounty',
  ADAMANT_ICO: 'ADAMANT Tokens'
}

export const BTC_BASED = Object.freeze([Cryptos.DOGE, Cryptos.DASH, Cryptos.BTC])

export const LSK_BASED = Object.freeze([Cryptos.LSK])

export const INSTANT_SEND = Object.freeze([Cryptos.DASH])

// Some cryptos allows to save public data with a Tx
export const ALLOW_TEXT_DATA = Object.freeze([Cryptos.LSK])

export const isErc20 = (crypto) => CryptosInfo[crypto].type === 'ERC20'

export const isEthBased = (crypto) => isErc20(crypto) || crypto === Cryptos.ETH

export const isFeeEstimate = (crypto) => isEthBased(crypto)

export const isBtcBased = (crypto) => BTC_BASED.includes(crypto)

export const isLskBased = (crypto) => LSK_BASED.includes(crypto)

export const isSelfTxAllowed = (crypto) => LSK_BASED.includes(crypto) || crypto === Cryptos.ADM

export const isInstantSendPossible = (crypto) => INSTANT_SEND.includes(crypto)

export const isTextDataAllowed = (crypto) => ALLOW_TEXT_DATA.includes(crypto)

export const RE_LSK_ADDRESS_LEGACY = /^[0-9]{2,21}L$/

/**
 * These gas limit values are used only for estimate fees for ETH & ERC-20 transfers in the Send tokens form,
 * Actual gas limit values are calculated with estimateGas(transactionObject)
 * when each specific transaction is created
 */

/** Gas limit value for the ETH transfers */
export const DEFAULT_ETH_TRANSFER_GAS = CryptosInfo['ETH'].defaultGasLimit
/** Gas limit value for the ERC-20 transfers */
export const DEFAULT_ERC20_TRANSFER_GAS = DEFAULT_ETH_TRANSFER_GAS * 2.4

/** Increase fee multiplier. Used as a checkbox on SendFundsForm */
export const INCREASE_FEE_MULTIPLIER = 2

export { Cryptos, CryptosInfo, CryptosOrder }

export default {
  EPOCH,
  Transactions
}

export const UserPasswordArticleLink =
  'https://medium.com/adamant-im/more-convenience-login-to-the-web-messenger-with-user-password-9d48a736dfd8'

export const UserPasswordHashSettings = {
  SALT: 'salt',
  ITERATIONS: 100000,
  KEYLEN: 64,
  DIGEST: 'sha512'
}

/** Status of ADM or coin transaction */
export const TransactionStatus = {
  CONFIRMED: 'CONFIRMED', // Tx has at least 1 network confirmation
  PENDING: 'PENDING', // We don't know about this Tx anything yet, it may not exist in a blockchain
  REGISTERED: 'REGISTERED', // This Ts is seen on a blockchain, but has 0 confirmations yet
  REJECTED: 'REJECTED', // We unable to find this Tx in a blockchain, or it says it was rejected
  INVALID: 'INVALID', // Wrong sender, recipient, time or amount
  UNKNOWN: 'UNKNOWN' // We don't recognize a cryptocurrency
}

/** Additional status of ADM or coin transaction, which leads to new virtualStatus */
export const TransactionAdditionalStatus = {
  NONE: false,
  INSTANT_SEND: 'instant_send', // Dash InstantSend enabled transaction
  ADM_REGISTERED: 'adm_registered' // ADM tx, registered in a blockchain, but has 0 confirmations yet
}

export const tsIcon = function (status) {
  if (status === TransactionStatus.CONFIRMED) {
    return 'mdi-check'
  } else if (status === TransactionStatus.PENDING || status === TransactionStatus.REGISTERED) {
    return 'mdi-clock-outline'
  } else if (status === TransactionStatus.REJECTED) {
    return 'mdi-close-circle-outline'
  } else if (status === TransactionStatus.INVALID) {
    return 'mdi-alert-outline'
  } else if (status === TransactionStatus.UNKNOWN) {
    return 'mdi-help-circle-outline'
  }
}

export const tsColor = function (status) {
  if (status === TransactionStatus.REJECTED) {
    return 'red'
  } else if (status === TransactionStatus.INVALID || status === TransactionStatus.UNKNOWN) {
    return 'yellow'
  }
  return ''
}

export const tsUpdatable = function (status, currency) {
  currency = currency.toUpperCase()
  if (currency === Cryptos.ADM) {
    return false
  } else if (status === TransactionStatus.CONFIRMED) {
    return true
  } else if (status === TransactionStatus.PENDING || status === TransactionStatus.REGISTERED) {
    return false
  } else if (status === TransactionStatus.REJECTED) {
    return true
  } else if (status === TransactionStatus.INVALID) {
    return true
  } else if (status === TransactionStatus.UNKNOWN) {
    return false
  }
}

/**
 * Returns minimal amount that can be transferred for the specified crypto
 * @param {string} crypto crypto
 * @returns {number}
 */
export function getMinAmount(crypto) {
  let amount = CryptosInfo[crypto].minTransferAmount

  if (!amount) {
    const precision = CryptosInfo[crypto].cryptoTransferDecimals
    amount = precision ? Math.pow(10, -precision) : 0
  }

  return amount
}

/**
 * When clicking on replied message inside the chat that animation duration will
 * be applied while scrolling.
 **/
export const SCROLL_TO_REPLIED_MESSAGE_ANIMATION_DURATION = 300 // ms

export const FetchStatus = {
  Idle: 'idle',
  Loading: 'loading',
  Success: 'success',
  Error: 'error'
}

export const REACT_EMOJIS = {
  FACE_WITH_TEARS_OF_JOY: 'üòÇ',
  FIRE: 'üî•',
  GRINNING_FACE_WITH_SMILING_EYES: 'üòÑ',
  THUMB_UP: 'üëç',
  OK_HAND: 'üëå',
  RED_HEART: '‚ù§Ô∏è',
  SLIGHTLY_SMILING_FACE: 'üôÇ',
  THINKING_FACE: 'ü§î',
  WAVING_HAND: 'üëã',
  FOLDED_HANDS: 'üôè',
  FLUSHED_FACE: 'üò≥',
  PARTY_POPPER: 'üéâ'
}
