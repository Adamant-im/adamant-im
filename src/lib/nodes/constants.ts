export type TNodeLabel =
  | 'adm-node'
  | 'eth-node'
  | 'eth-indexer'
  | 'btc-node'
  | 'btc-indexer'
  | 'doge-node'
  | 'doge-indexer'
  | 'dash-node'
  | 'ipfs-node'
  | 'kly-node'
  | 'kly-indexer'
  | 'rates-info'

type KebabToCamelCase<S extends string> = S extends `${infer T}-${infer U}`
  ? `${T}${Capitalize<KebabToCamelCase<U>>}`
  : S

type KebabToPascalCase<S extends string> = Capitalize<KebabToCamelCase<S>>

type CamelToKebabCase<S extends string> = S extends `${infer T}${infer U}`
  ? `${T extends Capitalize<T> ? '-' : ''}${Lowercase<T>}${CamelToKebabCase<U>}`
  : S

type PascalToKebabCase<S extends string> = CamelToKebabCase<Uncapitalize<S>>

type NodeLabels<V extends TNodeLabel = TNodeLabel> = {
  [K in KebabToPascalCase<V>]: Extract<V, PascalToKebabCase<K>>
}

export const NODE_LABELS: NodeLabels = {
  AdmNode: 'adm-node',
  EthNode: 'eth-node',
  EthIndexer: 'eth-indexer',
  BtcNode: 'btc-node',
  BtcIndexer: 'btc-indexer',
  DogeNode: 'doge-node',
  DogeIndexer: 'doge-indexer',
  DashNode: 'dash-node',
  IpfsNode: 'ipfs-node',
  KlyNode: 'kly-node',
  KlyIndexer: 'kly-indexer',
  RatesInfo: 'rates-info'
}
