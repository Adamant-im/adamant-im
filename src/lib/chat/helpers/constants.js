import { Cryptos } from '@/lib/constants'

/**
 * Maps crypto transfer transaction types to the respective cryptos.
 * E.g.:
 *
 * ```
 * {
 *   'eth_transaction': 'ETH',
 *   'bnb_transaction': 'BNB'
 * }
 * ```
 */
export const KnownCryptos = Object.keys(Cryptos).reduce((map, crypto) => {
  if (crypto !== Cryptos.ADM) {
    const key = `${crypto.toLowerCase()}_transaction`
    map[key] = crypto
  }
  return map
}, {})

/** Cryptos, supported by other clients, but not PWA */
export const UnsupportedCryptos = {
  lsk_transaction: 'LSK'
}
