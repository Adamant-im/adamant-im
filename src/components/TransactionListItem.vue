<template>
  <div :class="className">
    <v-list-item avatar :class="`${className}__tile`" @click="onClickTransaction">
      <template #prepend>
        <v-icon
          :class="`${className}__prepend-icon`"
          :icon="
            isStringEqualCI(senderId, userId) ? mdiAirplaneTakeoff : mdiAirplaneLanding
          "
          size="small"
        />
      </template>

      <v-list-item-title v-if="partnerName">
        <span class="a-text-regular-enlarged">{{ partnerName }}</span>
        <span class="a-text-explanation-enlarged"> ({{ partnerId }})</span>
      </v-list-item-title>
      <v-list-item-title v-else>
        <span class="a-text-regular-enlarged">{{ partnerId }}</span>
      </v-list-item-title>

      <v-list-item-title>
        <span :class="`${className}__amount ${directionClass}`">{{
          currency(amount, crypto)
        }}</span>
        <span :class="`${className}__rates`">{{ historyRate }}</span>
        <span v-if="comment" class="a-text-regular-enlarged-bold" style="font-style: italic">
          "</span
        >
        <span v-if="comment" class="a-text-explanation" style="font-weight: 100">{{
          comment
        }}</span>
        <span v-if="textData" class="a-text-regular-enlarged-bold" style="font-style: italic">
          #</span
        >
        <span v-if="textData" class="a-text-explanation" style="font-weight: 100">{{
          textData
        }}</span>
      </v-list-item-title>

      <v-list-item-subtitle :class="`${className}__date`" class="a-text-explanation-small">
        <span v-if="!isPendingTransaction">{{ formatDate(createdAt) }}</span>
        <span v-else-if="status" :class="`${className}__status`">{{
          $t(`transaction.statuses.${status}`)
        }}</span>
      </v-list-item-subtitle>

      <template #append>
        <v-list-item-action v-if="isClickIcon" :class="`${className}__action`" end>
          <v-btn icon ripple variant="plain" @click.stop="onClickIcon">
            <v-icon
              :class="`${className}__icon`"
              :icon="isPartnerInChatList ? mdiMessageText : mdiMessageOutline"
              size="small"
            />
          </v-btn>
        </v-list-item-action>
      </template>
    </v-list-item>

    <v-divider :inset="true" class="a-divider" />
  </div>
</template>

<script>
import formatDate from '@/filters/date'
import { EPOCH, Cryptos, TransactionStatus } from '@/lib/constants'
import partnerName from '@/mixins/partnerName'
import { isStringEqualCI } from '@/lib/textHelpers'
import currencyAmount from '@/filters/currencyAmount'
import { timestampInSec } from '@/filters/helpers'
import currency from '@/filters/currencyAmountWithSymbol'
import { mdiAirplaneLanding, mdiAirplaneTakeoff, mdiMessageOutline, mdiMessageText } from '@mdi/js'


export default {
  mixins: [partnerName],
  props: {
    id: {
      type: String,
      required: true
    },
    // Crypto address, like 1F9bMGsui6GbcFaGSNao5YcjnEk38eXXg7 or U3716604363012166999
    senderId: {
      type: String,
      required: true
    },
    status: {
      type: String,
      required: true
    },
    recipientId: {
      type: String,
      required: true
    },
    textData: {
      type: String,
      required: false
    },
    timestamp: {
      type: Number,
      required: true
    },
    amount: {
      type: [Number, String],
      required: true
    },
    crypto: {
      type: String,
      default: 'ADM',
      validator: (v) => v in Cryptos
    }
  },
  emits: ['click:transaction', 'click:icon'],
  setup() {
    return {
      mdiAirplaneLanding,
      mdiAirplaneTakeoff,
      mdiMessageOutline,
      mdiMessageText
    }
  },
  data: () => ({
    virtualTimestamp: Date.now()
  }),
  computed: {
    // Own crypto address, like 1F9bMGsui6GbcFaGSNao5YcjnEk38eXXg7 or U3716604363012166999
    userId() {
      if (this.crypto === Cryptos.ADM) {
        return this.$store.state.address
      } else {
        const cryptoModule = this.crypto.toLowerCase()
        return this.$store.state[cryptoModule].address
      }
    },
    // Crypto address, like 1F9bMGsui6GbcFaGSNao5YcjnEk38eXXg7 or U3716604363012166999
    partnerId() {
      return isStringEqualCI(this.senderId, this.userId) ? this.recipientId : this.senderId
    },
    // Partner's ADM address, if found. Else, returns 'undefined'
    partnerAdmId() {
      const admTx = this.getAdmTx
      return isStringEqualCI(admTx.senderId, this.$store.state.address)
        ? admTx.recipientId
        : admTx.senderId
    },
    // Partner's name from KVS, if partnerAdmId found
    partnerName() {
      if (isStringEqualCI(this.partnerId, this.userId)) {
        return this.$t('transaction.me')
      }
      const name = this.getPartnerName(this.partnerAdmId) || ''
      if (this.isCryptoADM()) {
        return name
      } else {
        return name || this.partnerAdmId || ''
      }
    },
    createdAt() {
      if (this.crypto === 'ADM') {
        return this.timestamp * 1000 + EPOCH
      }
      return this.timestamp
    },
    isPartnerInChatList() {
      return this.$store.getters['chat/isPartnerInChatList'](this.partnerAdmId)
    },
    className() {
      return 'transaction-item'
    },
    isClickIcon() {
      const hasPartnerAddress = this.partnerAdmId && this.partnerAdmId !== undefined
      return this.isCryptoADM() || hasPartnerAddress
    },
    getAdmTx() {
      return this.admTx()
    },
    directionClass() {
      if (
        isStringEqualCI(this.senderId, this.userId) &&
        isStringEqualCI(this.recipientId, this.userId)
      ) {
        return `${this.className}__amount--is-itself`
      } else if (isStringEqualCI(this.senderId, this.userId)) {
        return `${this.className}__amount--is-outgoing`
      } else {
        return `${this.className}__amount--is-incoming`
      }
    },
    comment() {
      const admTx = this.getAdmTx
      return admTx.message
    },
    historyRate() {
      const amount = currencyAmount(this.amount, this.crypto)
      return (
        '~' +
        this.$store.getters['rate/historyRate'](
          timestampInSec(this.crypto, this.timestamp || this.virtualTimestamp),
          amount,
          this.crypto
        )
      )
    },
    isPendingTransaction() {
      return (
        this.status === TransactionStatus.PENDING || this.status === TransactionStatus.REGISTERED
      )
    }
  },
  mounted() {
    this.getHistoryRates()
  },
  methods: {
    isStringEqualCI(string1, string2) {
      return isStringEqualCI(string1, string2)
    },
    isCryptoADM() {
      return this.crypto === Cryptos.ADM
    },
    onClickTransaction() {
      this.$emit('click:transaction', this.id)
    },
    onClickIcon() {
      this.$emit('click:icon', this.partnerAdmId)
    },
    admTx() {
      // Note: because to Chatrooms API, we store only fetched messages, not all of them
      // So, we'll search through only stored messages

      if (this.isCryptoADM()) {
        const chatWithPartner = this.$store.state.chat.chats[this.partnerId]
        const msg =
          chatWithPartner && chatWithPartner.messages
            ? chatWithPartner.messages.find((message) => message.id === this.id)
            : undefined
        return (
          // this.$store.getters['chat/partnerMessageById'](this.partnerId, this.id) || // unable to use getter, as it causes "You may have an infinite update loop in a component render function"
          msg || // first, try to get message with comment
          this.$store.state.adm.transactions[this.id] || // next, try to get direct transaction
          {} // finally, return empty object
        )
      }

      // If crytpo is not ADM: we have to scan all the messages
      const admTx = {}
      Object.values(this.$store.state.chat.chats).some((chat) => {
        Object.values(chat.messages).some((msg) => {
          if (msg.hash && msg.hash === this.id) {
            Object.assign(admTx, msg)
          }
          return !!admTx.id
        })
        return !!admTx.id
      })
      return admTx
    },
    getHistoryRates() {
      this.$store.dispatch('rate/getHistoryRates', {
        timestamp: timestampInSec(this.crypto, this.timestamp || this.virtualTimestamp)
      })
    },
    currency,
    formatDate
  }
}
</script>

<style lang="scss" scoped>
@import '@/assets/styles/themes/adamant/_mixins.scss';
@import '@/assets/styles/settings/_colors.scss';

.transaction-item {
  &__rates {
    color: hsla(0, 0%, 100%, 0.7);
    font-style: italic;
    @include a-text-regular();
    margin-left: 4px;
  }
  &__amount {
    @include a-text-regular-enlarged-bold();
  }
  &__date {
    margin-top: 4px;
  }
  &__prepend-icon {
    margin-inline-end: 16px;
    margin-top: 8px;
  }
  :deep(.v-divider--inset:not(.v-divider--vertical)) {
    margin-left: 56px;
    max-width: calc(100% - 56px);
  }
  &__action {
    min-width: 36px;
  }
  &__status {
    color: map-get($adm-colors, 'attention');
  }
  // Do not break computed length of v-divider
  /*&__tile*/
  /*:deep(.v-list__tile)*/
  /*padding: 0 12px*/
}

/** Themes **/
.v-theme--light.v-list {
  .transaction-item {
    &__amount {
      color: map-get($adm-colors, 'regular');
    }
    &__rates {
      color: map-get($adm-colors, 'muted');
      &--is-incoming {
        color: map-get($adm-colors, 'good');
      }
      &--is-outgoing {
        color: map-get($adm-colors, 'danger');
      }
    }
    &__icon {
      color: map-get($adm-colors, 'muted');
    }
  }
}
.v-theme--dark.v-list {
  .transaction-item {
    &__amount {
      color: map-get($adm-colors, 'regular');
      &--is-incoming {
        color: map-get($adm-colors, 'good');
      }
      &--is-outgoing {
        color: map-get($adm-colors, 'danger');
      }
    }
  }
}
</style>
