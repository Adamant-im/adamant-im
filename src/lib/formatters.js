import { EPOCH } from './constants'

function getTime (date) {
  const hours = date.getHours()
  let time = ''
  if (hours < 10) {
    time = '0' + hours
  } else {
    time = '' + hours
  }
  time = time + ':'
  const minutes = date.getMinutes()
  if (minutes < 10) {
    time = time + '0' + minutes
  } else {
    time = time + '' + minutes
  }
  return time
}

function formatDate (timestamp) {
  timestamp = parseInt(timestamp)
  // That's for the ADM timestamps, which use EPOCH as a base.
  // Other cryptos use normal timestamps
  if (timestamp < EPOCH) {
    timestamp = timestamp * 1000 + EPOCH
  }

  const startToday = new Date()
  startToday.setHours(0, 0, 0, 0)

  const date = new Date(timestamp)
  if (date.getTime() > startToday.getTime()) {
    return this.$t('chats.date_today') + ', ' + getTime(date)
  }

  const startYesterday = new Date(startToday.getTime() - 86400000)
  if (date.getTime() > startYesterday.getTime()) {
    return this.$t('chats.date_yesterday') + ', ' + getTime(date)
  }

  let options = { weekday: 'short' }
  if ((Date.now() - timestamp) > (4 * 3600 * 24 * 1000)) {
    options = { day: 'numeric', month: 'short' }
  }

  if (startToday.getFullYear() !== date.getFullYear()) {
    options.year = 'numeric'
  }

  return date.toLocaleDateString(this.$t('region'), options) + ', ' + getTime(date)
}

const plugin = {
  install (app) {
    app.config.globalProperties.$formatDate = formatDate
  }
}

export default plugin
