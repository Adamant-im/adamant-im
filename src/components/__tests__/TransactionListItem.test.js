import { shallowMount } from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import Vuetify from 'vuetify'

import currencyFilter from '@/filters/currency'
import { EPOCH } from '@/lib/constants'
import formatters from '@/lib/formatters'
import mockupI18n from './__mocks__/plugins/i18n'
import TransactionListItem from '@/components/TransactionListItem'

Vue.filter('currency', currencyFilter)
Vue.use(Vuex)
Vue.use(VueI18n)
Vue.use(Vuetify)

// set $formatAmount & $formatDate prototype methods
formatters(Vue)

let fake = createFakeVars()

function createFakeVars () {
  return {
    partnerName: 'Rick',
    isPartnerInChatList: false
  }
}

/**
 * Mockup store helper.
 */
function mockupStore () {
  const chat = {
    getters: {
      isPartnerInChatList: () => () => fake.isPartnerInChatList
    },
    namespaced: true
  }

  const partners = {
    getters: {
      displayName: () => () => fake.partnerName
    },
    namespaced: true
  }

  const store = new Vuex.Store({
    modules: {
      chat,
      partners
    }
  })

  return {
    store,
    chat,
    partners
  }
}

/**
 * Mockup valid props.
 */
const validProps = {
  id: '123456789',
  senderId: 'U654321',
  recipientId: 'U123456',
  timestamp: 37835483, // ADM timestamp
  amount: 100000000 // 1 ADM
}

// disable Vue console warnings when testings
// with invalid `propsData`
// jest.spyOn(global.console, 'error').mockImplementation(() => jest.fn())

describe('TransactionListItem', () => {
  let i18n = null
  let store = null
  let partners = null // vuex module
  let chat = null // vuex module

  beforeEach(() => {
    const vuex = mockupStore()

    store = vuex.store
    partners = vuex.partners
    chat = vuex.chat

    i18n = mockupI18n()
    fake = createFakeVars()
  })

  // it('renders the correct markup', () => {
  //   const wrapper = shallowMount(TransactionListItem, {
  //     store,
  //     i18n,
  //     propsData: {
  //       ...validProps
  //     }
  //   })
  //
  //   expect(wrapper.element).toMatchSnapshot()
  // })

  it('should return computed.partnerName', () => {
    const wrapper = shallowMount(TransactionListItem, {
      store,
      i18n,
      propsData: {
        ...validProps
      }
    })

    expect(wrapper.vm.partnerName).toBe('Rick')
  })

  it('should return correct date based on props.timestamp', () => {
    const admTimestamp1 = 3600 * 24 // 1 day
    const admTimestamp2 = 3600 * 24 * 2 // 2 days

    const wrapper = shallowMount(TransactionListItem, {
      store,
      i18n,
      propsData: {
        ...validProps,
        timestamp: admTimestamp1
      }
    })

    // Sep 3, 2017, 20:00
    expect(wrapper.vm.createdAt).toBe(admTimestamp1 * 1000 + EPOCH)

    // Sep 4, 2017, 20:00
    wrapper.setProps({ timestamp: 3600 * 24 * 2 })
    expect(wrapper.vm.createdAt).toBe(admTimestamp2 * 1000 + EPOCH)
  })

  it('should not display icon when transaction contains a message', () => {
    const wrapper = shallowMount(TransactionListItem, {
      store,
      i18n,
      propsData: {
        ...validProps
      }
    })

    // transaction without comment
    expect(wrapper.vm.isPartnerInChatList).toBe(false)
  })

  it('should display icon when transaction contains a message', () => {
    fake.isPartnerInChatList = true

    const wrapper = shallowMount(TransactionListItem, {
      store,
      i18n,
      propsData: {
        ...validProps
      }
    })

    expect(wrapper.vm.isPartnerInChatList).toBe(true)
  })

  it('should emit events', () => {
    const wrapper = shallowMount(TransactionListItem, {
      store,
      i18n,
      propsData: {
        ...validProps
      }
    })

    wrapper.vm.onClickTransaction()
    wrapper.vm.onClickIcon()

    expect(wrapper.emitted()['click:transaction']).toEqual([[wrapper.vm.id]]) // emit with transactionId
    expect(wrapper.emitted()['click:icon']).toBeTruthy()
  })
})
