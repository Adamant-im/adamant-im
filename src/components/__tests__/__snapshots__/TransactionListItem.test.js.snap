// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionListItem renders the correct markup 1`] = `
<v-list-tile-stub
  activeclass="primary--text"
  avatar="true"
>
  <v-list-tile-avatar-stub
    size="40"
  >
    <v-icon-stub>
      mdi-airplane-takeoff
    </v-icon-stub>
  </v-list-tile-avatar-stub>
   
  <v-list-tile-content-stub>
    <v-list-tile-title-stub>
      
      Rick
      
      <span
        class="body-1 grey--text text--darken-1"
      >
        (U654321)
      </span>
    </v-list-tile-title-stub>
     
    <v-list-tile-sub-title-stub>
      <v-layout-stub
        tag="div"
      >
        
        1 ADM
        
        <v-spacer-stub />
        
        14.11.2018
      
      </v-layout-stub>
    </v-list-tile-sub-title-stub>
  </v-list-tile-content-stub>
   
  <!---->
</v-list-tile-stub>
`;
